### ISSUE TITLE
support Nostr

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#446]
Simple, key-centric, UX is still rough, community is heavily Bitcoin/crypto people right now, but could be promising.

https://github.com/nostr-protocol/nostr

**Comment by snarfed** [comment]
Data points to help determine priority:

- https://nostr.io/#profile-chart says there are 365k users (profiles) right now
- https://stats.nostr.band/ says there are 1.13M users (profiles) right now, 13k ["high" (ie non-zero) quality](https://trust.nostr.band/) WAUs

**Comment by danielcharrua** [comment]
I think they differ because of when they started collecting stats, maybe also the relays they are using. Here are two more:

- https://primal.net/explore
- https://nashboard.space

Talking to a nostr friend, he told me that https://stats.nostr.band/ is the most reliable source.

**Comment by snarfed** [comment]
Thank you!

**Comment by snarfed** [comment]
ActivityPub <=> Nostr bridge! https://soapbox.pub/blog/mostr-fediverse-nostr-bridge/

**Comment by sven-hash** [comment]
Is there something to test already ?

**Comment by snarfed** [comment]
@sven-hash not here yet, no, but PRs are welcome!

**Comment by sven-hash** [comment]
Thanks @snarfed for the quick answer, good to know that PRs are welcome !
Will check on it with other Nostr people

**Comment by snarfed** [comment]
I reread https://github.com/nostr-protocol and https://github.com/nostr-protocol/nips/blob/master/01.md and have a few questions about the protocol:

- Should Bridgy Fed be a relay? Or should it act more like a client proxy, and use other relays? Not sure yet.
- Apart from events you've sent, it sounds like all of your account/user data, notably your private key and your follows (?), are kept locally on the client. Key rotation sounds unsolved, ok, fine - https://github.com/nostr-protocol/nostr/issues/45 - ~but how about your follows? If you use multiple devices, do you have to manually follow/unfollow on all devices to keep them in sync? If you use a single device, and you lose it, is your following list gone? Is there some NIP for this?~
  Aha looks like this is [NIP-02](https://github.com/nostr-protocol/nips/blob/master/02.md).
- Where's the guidance on whether/how long a relay should store a given event? Eg the contacts kind for NIP-02? Haven't found anything on that in NIP-01.

Here are the NIPS for common interactions I've found so far:
  - Likes/reactions: [NIP-25](https://github.com/nostr-protocol/nips/blob/master/25.md)
  - User mentions: [NIP-27](https://github.com/nostr-protocol/nips/blob/master/27.md)
  - Replies: [NIP-10](https://github.com/nostr-protocol/nips/blob/master/10.md), notably the _Marked "e" tags_ method, which has parent, root, and references (eg quotes), along _"p" tags_ for those events' authors and other related users. Sounds like clients can use this to do notifications, by querying for themselves in "p" tags?
  - Reposts: [NIP-18](https://github.com/nostr-protocol/nips/blob/master/18.md)
  - Hashtags: [NIP-12](https://github.com/nostr-protocol/nips/blob/master/12.md)
  - Articles: [NIP-23](https://github.com/nostr-protocol/nips/blob/master/23.md)

cc @manton

**Comment by snarfed** [comment]
To relay, or not to relay? Reasons:

Relay:
- ~~need to be a WebSocket server but not client, which may be harder on GAE~~
- necessary for discovery? ie for a Nostr user discovering another user, both of whom are new to Bridgy Fed, if we're not a relay, there's no way for BF to discover this and propagate events into the remote relay on demand. This might be the deciding factor.

Not relay:

- drastically less work, don't have to implement relay protocol, just be a client

Same either way:
- need to connect to an arbitrary number of external relays either way? to collect Nostr users' posts? Or would we say they have to add us as a relay destination for their posts to federate them?
  - How does relay-to-relay traffic work anyway? Nothing about it in NIP-01. Negentropy synching? How widespread is it?

**Comment by snarfed** [comment]
Interestingly Nostr [user growth](https://stats.nostr.band/#daily_new_users) [and posts](https://stats.nostr.band/#daily_events) seem to be _way_ down over the last month or so. Either that or maybe some big relay change happened and stats.nostr.band missed it. But it's "the most reliable source" ^ so 🤷

![image](https://github.com/snarfed/bridgy-fed/assets/778068/6e8bb9cb-8715-48cb-b7fe-a99a9c734698)

![image](https://github.com/snarfed/bridgy-fed/assets/778068/8243adac-71b1-4297-a4c2-e7e4f1251296)

...ah, ok, the earlier high new user growth was evidently accounts with no profile or other info, so they likely weren't human, so the drop in posts may have just been a big wave of spam ending.

Still, [active "trusted" users](https://stats.nostr.band/#daily_active_users) is slowly dropping too.

![image](https://github.com/snarfed/bridgy-fed/assets/778068/92d80693-37f8-4f5d-b40c-695e3ab4994e)

**Comment by snarfed** [comment]
List of relays: https://nostr.info/relays/

I'm currently testing with https://relayable.org/

**Comment by manton** [comment]
Curious what relays you settle on. I'm currently using nos.lol, relay.snort.social, relay.damus.io, and relay.nostrgraph.net because they seemed popular and didn't sound terrible. But I'd like to put some more thought into this. There's a performance trade-off to supporting "all" the relays.

**Comment by snarfed** [comment]
Definitely. Right now I'm just learning, but I suspect Bridgy Fed will actually need to _be_ a relay itself: https://github.com/snarfed/bridgy-fed/issues/446#issuecomment-1563618821 . Not sure yet though, and obviously it's pretty different from micro.blog.

**Comment by snarfed** [comment]
I recently finished enough Nostr data conversion and relay protocol support to launch it in granary:
* https://granary.io/?site=nostr#logins
* https://granary.io/?input=html&output=nostr#logins

**Comment by snarfed** [comment]
Current obstacle: [from NIP-01](https://github.com/nostr-protocol/nips/blob/master/01.md):

> Each user has a keypair. Signatures, public key, and encodings are done according to the [Schnorr signatures standard for the curve secp256k1](https://bips.xyz/340).

I'm having trouble finding a production-appropriate Python implementation of [BIP340](https://bips.xyz/340) Schnorr signatures.

* [pycryptodome feature request](https://github.com/Legrandin/pycryptodome/issues/762)
* [pyca/cryptography feature request](https://github.com/pyca/cryptography/issues/8202) (closed pending OpenSSL support)
* [OpenSSL feature request](https://github.com/openssl/openssl/issues/8440)
* BIP340 includes a [reference Python implementation](https://github.com/bitcoin/bips/blob/master/bip-0340/reference.py), but it's unpackaged and unmaintained
* [BitPolito/schnorr-sig](https://github.com/BitPolito/schnorr-sig) is "for educational purposes only"
* [btclib](https://btclib.org/) looks mature but says it's "not intended for production environments"
* [FROST-BIP340](https://github.com/jesseposner/FROST-BIP340) is an early implementation of a new, BIP340-compatible multi-sig schema
* [long discussion on StackExchange](https://crypto.stackexchange.com/questions/34863/ec-schnorr-signature-multiple-standard)

**Comment by snarfed** [comment]
Production-ready candidates:
* [rustyrussell/secp256k1-py](https://github.com/rustyrussell/secp256k1-py), FFI bindings around [bitcoin-core/secp256k1](https://github.com/bitcoin-core/secp256k1)
* [Moustikitos/fast-curve](https://github.com/Moustikitos/fast-curve), native lib, seems largely unused
* [karask/python-bitcoin-utils](https://github.com/karask/python-bitcoin-utils), just copied the reference implementation with minor modifications
* [kauriid/schnorrpy](https://gitlab.com/kauriid/schnorrpy), bindings for the Rust schnorrkel lib, looks unmaintained, unclear if it does full BIP340

secp256k1-py looks like the winner.

**Comment by snarfed** [comment]
To try: https://www.nos.social/

**Comment by snarfed** [comment]
I should do a survey of relay info metadata, [specifically the `limitation` field](https://github.com/nostr-protocol/nips/blob/master/11.md#server-limitations), for a handful of the bigger relays. nostr.law (?), nostr.mom, relayable.org, etc. I'm curious what their limits on the various filter fields are, esp if I need to subscribe for all Bridgy Fed users on a given relay.

**Comment by snarfed** [comment]
Started a todo list in the description

**Comment by snarfed** [comment]
Active user counts are still low, but trending up again.

https://stats.nostr.band/#daily_active_users
<img width="1135" alt="image" src="https://github.com/snarfed/bridgy-fed/assets/778068/2a1d7560-222e-4cb8-b3ed-406882f9ad8b">

https://stats.nostr.band/#weekly_active_users
<img width="1137" alt="image" src="https://github.com/snarfed/bridgy-fed/assets/778068/2abd81d5-04ee-452c-be95-fe68c386954d">

**Comment by snarfed** [comment]
6 mos later, WAUs have been flat, DAUs flat to slightly down. 🤷 

![image](https://github.com/user-attachments/assets/22bf7501-bb9d-4db8-8dfb-f6374c7c6390)

![image](https://github.com/user-attachments/assets/a2b26468-1615-4ae9-971e-b0ed3f6e8cf0)

**Comment by snarfed** [comment]
Got some great tips on this from @rabble (thanks man!) at ATmosphere this past weekend:

* Use NIP-05, both publishing and validating (bidirectional!). Lots of clients (and relays?) use it as a spam signal.
* Use NIP-42, authed connection to relays. Some of them require it for writes.
* Validate sigs! Lots of spammers etc copy other people's profiles, posts, NIP-05s, and try to pass them off as their own.
* Ask him for a good short list of starting relays to use.

**Comment by snarfed** [comment]
Re secp256k1, [lots more candidates on pypi](https://pypi.org/search/?q=secp256k1), eg https://github.com/scgbckbone/python-secp256k1.

**Comment by snarfed** [comment]
- Argh. I started using https://github.com/rustyrussell/secp256k1-py , which works great, _except_ that it seems to [only support 33 and 65 byte public keys](https://github.com/rustyrussell/secp256k1-py/issues/4), not 32 byte, [which NIP-19 specifies](https://nips.nostr.com/19#bare-keys-and-ids). 😕

**Comment by snarfed** [comment]
Looks like maybe the extra byte is always one of two prefixes, `0x02` or `0x03`, and for keypairs generated by secp256k1-py, the difference doesn't actually matter. I can replace the `0x03` with `0x02`, and it still happily loads the resulting public key and verifies signatures generated with the original `0x03` version, and vice versa...and it works with keypairs generated elsewhere too.

**Comment by melvincarvalho** [comment]
Taproot chops off the 02 / 03 which is the Y coordinate.  It's loosely assumed to be 02 for the purposes of schnorr signing, but in reality the full key would be positive or negative, which is the older compressed format.  In nostr the 02 / 03 is chopped off too, leaving the x-only coordinate on the curve.

**Comment by snarfed** [comment]
Re NIP-42 relay authentication, asked https://github.com/nostr-protocol/nips/issues/1949

**Comment by snarfed** [comment]
A few candidates for relays to send our outbound events to, largely chosen just based on size, from https://nostr.watch/relays and https://nostr.info/relays/ :

* https://nos.lol, runs strfry. [Some info here](https://wikifreedia.xyz/a/naddr1qvzqqqrcvgpzp8lvwt2hnw42wu40nec7vw949ys4wgdvums0svs8yhktl8mhlpd3qy88wumn8ghj7mn0wvhxcmmv9uqqwmn0wvhxcmmvxawj0s); [can't find much else](https://www.google.com/search?q="nos.lol").
* https://nostr.mom, runs strfry. [Some info here](https://wikifreedia.xyz/a/naddr1qvzqqqrcvgpzp8lvwt2hnw42wu40nec7vw949ys4wgdvums0svs8yhktl8mhlpd3qy88wumn8ghj7mn0wvhxcmmv9uqqjmn0wd68ytnddakspvfv36); [can't find much else](https://www.google.com/search?q="nostr.mom").
* ~~https://nostr.wine, "the most reliable paid relay on nostr." Runs strfry. [Claims to aggregate events from dozens of other relays, and broadcast created events back out to multiple other relays.](https://docs.nostr.wine/filter/readme#how-does-it-work) I don't understand their [pricing](https://nostr.wine/add-time) though; you buy "days," but I don't know what that means.~~ It's paid per user, 10k sats ie $10/mo. Nope.
* https://relay.damus.io, by the Damus people (guy?), runs strfry, not a ton of other info though
* https://relay.nostr.band, seems promising

https://github.com/PrimalHQ/primal-server/blob/main/default-relays.json also looks like a decent list.

https://stats.nostr.band/#relay_users has a few charts of the biggest/most popular relays in various ways, eg:

<img width="1167" alt="Image" src="https://github.com/user-attachments/assets/ac32fb8b-28b9-4c6e-bf32-a0833c295240" />

@rabble thoughts?

**Comment by fiatjaf** [comment]
Some comments on **being a relay**: you don't have to fully implement a relay, you can run [strfry](https://github.com/hoytech/strfry), which is the most popular implementation out there, very fast and efficient, and it has a plugin system (strfry communicates with an external process via stdin/stdout) that allows you to deny any events that aren't coming from the bridge, it's easy.

Why would you do it?

 - it's cleaner and more beautiful (or at least I think).
 - you don't have to depend on third parties that might reject your events (and the free ones might start blocking you if you flood them too much) or go offline etc.
 - you can query your own events easily, locally (there is an `strfry export` command that takes a JSON filter and is better than opening a connection if you're in the same machine) -- if for whatever reason you happen to need such a thing.
 - would allow for directly browsing that specific relay (a flow that is nascent on Nostr right now, supported in some clients, see for example https://jumble.social/?r=pyramid.fiatjaf.com) if anyone wants to "see what bridged accounts are saying" (they would just load the bridgy relay feed directly). I've written about this on https://fiatjaf.com/d36862b9.html#the-wss-nostr-zbd-gg-relay-637 when I was faced with the same choice when building something like a "bridge" myself  too (not really a bridge, but a custodial server for users of a closed platform to publish to and read from Nostr).

Why not?

 - because you don't care about any of the above.
 - see my next comment for more reasons.

**Comment by fiatjaf** [comment]
Some comments on **relay selection**:

The basic Nostr answer to the question "where do I find Bob's posts?" is "on the relays Bob publishes to". Historically, though, because it wasn't clear how do to that in practice (and because of a bad precedent I set with my first client) many clients were implemented in a naïve way that just connected to a static set of relays and tried to fetch all posts they needed from that static set.

Then a pattern emerged that became known as "gossip model" (because of the client named Gossip, not because it had anything to do with actual gossiping) and later "outbox model". Well, the pattern is not really a pattern, it's more like a goal.

In short, NIP-65 was written and the basic idea is that you consult some well-known (basically hardcoded) relays for the `kind:10002` event of any given user -- which contains a list of relays that user has announced to be publishing at --, then you look into those relays for everything else from that person. If this is not clear already maybe this will help (or maybe not): https://how-nostr-works.pages.dev/#/outbox. (and there are mitigation possibilities for the cases in which these "well known" relays is bad for whatever reason, but we don't have to concern ourselves with that for now I think).

So this is basically the "correct" way of fetching posts from people that _most_ clients do. On the other hand, the few _most popular_ (namely Amethyst, Damus and Primal) clients still do not do it (although they have all declared that they will start doing it in a way or another anytime soon). Which means that these clients miss all notes that aren't published to the top 5 most famous relays (more or less, because users can edit the static list, but that's an ugly workaround and probably very few do -- and Primal is a peculiar anomaly, we can talk about it later).

Back to **the question of Bridgy running its own relay**:

In other words: if you publish bridged posts only to a dedicated Bridgy relay that will be:

  - great and glorious, as it will act as an extra incentive for these clients to fix their stuff.
  - not great, because a bunch of users won't see them.

Well, you can always publish to a Bridgy relay and another relay. Or start doing one then change afterwards.

**Comment by fiatjaf** [comment]
Now, with regards to **where Bridgy should go to fetch posts from people** that it wants to bridge to external networks, I think it should do the basic NIP-65 flow (fetch `kind:10002`) then connect to these. It can be complicated to do that from a single server trying to fetch posts from a huge amount of Nostr users -- I know because that's what I did on ZBD Social, but what is the fun if it's not hard?

---

Oh, another question you have up there somewhere in this topic is something like: **if a bridged user gets a reply or a mention, where will that post be?** NIP-65 also has an answer to that: in a user's `kind:10002` there are "read" relays, which are to be used by the replier anytime it is replying to someone (or in any other case in which someone is being mentioned).

So "good" clients will do that and send replies to wherever you announce your bridged users' "read" relays are. But, again, because some clients are obtuse you might have to read from the top-5 big relays in order to get most of the replies and mentions regardless, which makes me sad.

**Comment by snarfed** [comment]
Thank you for all this @fiatjaf! Hugely helpful. I was definitely aware of NIP-65 and fully plan to use it, both inbound and outbound. And thanks also for the real world info. It's one thing to learn a network from its standards and reference code, quite another to learn the real world behavior and expectations. You, @rabble, @alexgleason, @manton and others have already been really helpful with that!

**Comment by melvincarvalho** [comment]
wss://ditto.pub/relay by @alexgleason might be a good choice too -- see https://soapbox.pub/ditto/

**Comment by snarfed** [comment]
Basic functionality is done and working! Posts, replies, reposts, likes, @-mentions, images, video, etc. Follows are still in progress, #2203.

Not launched publicly yet, but the todo list in this issue description has gotten unwieldy enough that I'm going to break out the remaining work into separate issues.

**Comment by snarfed** [comment]
stats.nostr.band is dead. 😢🫗

Haven't found a replacement I'm fully happy with yet, but https://stats.andotherstuff.org/ is maybe ok. Looks like (write) user counts right now are 21k MAU (Dec) 2k DAU (Feb). Oof.

**Comment by melvincarvalho** [comment]
Sounds about right, yes. Look at the w/ both for human numbers. I'm guessing lots of bots, too.

**Comment by TomCasavant** [comment]
diVine had a public launch a few weeks ago (https://techcrunch.com/2026/04/29/jack-dorsey-backed-vine-reboot-divine-launches-to-public/) and is built on top of nostr. I'm not sure if that has had a significant impact on nostr's MAU at all

**Comment by snarfed** [comment]
Yes! We talk with Rabble pretty regularly, we've been excited about diVine for a while. I don't know about user stats either, but we have rough early proof-of-concepts for bridging diVine accounts to both Bluesky and fediverse:
* https://bsky.app/profile/windowsxp-vines.divine.video.nostr.brid.gy
* https://mastodon.social/@windowsxp-vines.divine.video@nostr.brid.gy

**Comment by melvincarvalho** [comment]
Just adding this which has support for Solid, Nostr and ActivityPub, with BlueSky planned

https://jss.live/

**Comment by snarfed** [comment]
A Solid client! Whoa, those exist in the wild?! 😁

...ah, nm, it's a server.

**Comment by Sebastix** [comment]
Here are some other sources for metrics:

* https://nostrarchives.com/analytics
* https://ditto-metrics.shakespeare.wtf/
* https://npub.world/stats
